godot-ffi/src/binding/single_threaded.rs

```diff
@@ -0,0 +1,146 @@
+
+        Some(())
+    }
+
+    /// Get the binding from the binding storage.
+    ///
+    /// # Safety
+    /// - Must be called from the main thread.
+    /// - The binding must be initialized.
+    #[inline(always)]
+    pub unsafe fn get_binding_unchecked() -> &'static GodotBinding {
+        let storage = Self::storage();
+
+        if cfg!(debug_assertions) {
+            let main_thread = storage.main_thread.get().expect(
+                "Godot engine not available; make sure you are not calling it from unit/doc tests",
+            );
+            assert_eq!(main_thread, &std::thread::current().id(), "attempted to access binding from different thread than main thread; this is UB - use the \"experimental-threads\" feature.");
+            storage.binding.get().unwrap()
+        } else {
+            // SAFETY: This function can only be called when the binding is initialized and from the main thread, so we know that it's initialized.
+            unsafe { storage.binding.get().unwrap_unchecked() }
+        }
+    }
+
+    pub fn is_initialized() -> bool {
+        // SAFETY: We do not access `binding`.
+        let storage = unsafe { Self::storage() };
+        storage.main_thread.get().is_some()
+    }
+}
+
+// SAFETY: We ensure that `binding` is only ever accessed from the same thread that initialized it.
+unsafe impl Sync for BindingStorage {}
+// SAFETY: We ensure that `binding` is only ever accessed from the same thread that initialized it.
+unsafe impl Send for BindingStorage {}
+
+pub struct GodotBinding {
+    pub(super) interface: GDExtensionInterface,
+    pub(super) library: GDExtensionClassLibraryPtr,
+    pub(super) global_method_table: BuiltinLifecycleTable,
+    pub(super) class_server_method_table: OnceCell<ClassServersMethodTable>,
+    pub(super) class_scene_method_table: OnceCell<ClassSceneMethodTable>,
+    pub(super) class_editor_method_table: OnceCell<ClassEditorMethodTable>,
+    pub(super) builtin_method_table: OnceCell<BuiltinMethodTable>,
+    pub(super) utility_function_table: UtilityFunctionTable,
+    pub(super) runtime_metadata: GdextRuntimeMetadata,
+    pub(super) config: GdextConfig,
+}
+
+impl GodotBinding {
+    pub fn new(
+        interface: GDExtensionInterface,
+        library: crate::GDExtensionClassLibraryPtr,
+        global_method_table: BuiltinLifecycleTable,
+        utility_function_table: UtilityFunctionTable,
+        runtime_metadata: GdextRuntimeMetadata,
+        config: GdextConfig,
+    ) -> Self {
+        Self {
+            interface,
+            library: GDExtensionClassLibraryPtr(library),
+            global_method_table,
+            class_server_method_table: OnceCell::new(),
+            class_scene_method_table: OnceCell::new(),
+            class_editor_method_table: OnceCell::new(),
+            builtin_method_table: OnceCell::new(),
+            utility_function_table,
+            runtime_metadata,
+            config,
+        }
+    }
```

Comment:
Not clear why you qualify `GDExtensionClassLibraryPtr` in the parameter list, but not in the body 🙂

Reply:
Because it's a newtype local to this module, just there so i can implement `Sync` and `Send` for it (you cant implement `Sync` and `Send` for raw pointers). so the parameter list refers to the ffi-generated one (which is just a type alias for a raw pointer), and the body refers to the newtype.

Reply:
I see. I would not use the `GDExtension` prefix for types we define ourselves (except if we _have to_ keep parity, like in the 4.0-4.1 bridge). Maybe just `ClassLibraryPtr`?